util/topgen.py

```diff
@@ -1099,6 +981,145 @@ def main():
+    parser.add_argument('--dv-base-prefix',
+                        default='dv_base',
+                        help='Prefix for the DV register classes from which '
+                        'the register models are derived.')
+    # Generator options for compile time random netlist constants
+    parser.add_argument(
+        '--rnd_cnst_seed',
+        type=int,
+        metavar='<seed>',
+        help='Custom seed for RNG to compute netlist constants.')
+
+    args = parser.parse_args()
+
+    # check combinations
+    if args.top_ral:
+        args.no_top = True
+
+    if (args.no_top or args.no_xbar or
+            args.no_plic) and (args.top_only or args.xbar_only or
+                               args.plic_only or args.alert_handler_only):
+        log.error(
+            "'no' series options cannot be used with 'only' series options")
+        raise SystemExit(sys.exc_info()[1])
+
+    if not (args.top_ral or args.plic_only or args.alert_handler_only or
+            args.tpl):
+        log.error(
+            "Template file can be omitted only if '--hjson-only' is true")
+        raise SystemExit(sys.exc_info()[1])
+
+    if args.verbose:
+        log.basicConfig(format="%(levelname)s: %(message)s", level=log.DEBUG)
+    else:
+        log.basicConfig(format="%(levelname)s: %(message)s")
+
+    if not args.outdir:
+        outdir = Path(args.topcfg).parent / ".."
+        log.info("TOP directory not given. Use %s", (outdir))
+    elif not Path(args.outdir).is_dir():
+        log.error("'--outdir' should point to writable directory")
+        raise SystemExit(sys.exc_info()[1])
+    else:
+        outdir = Path(args.outdir)
+
+    out_path = Path(outdir)
+    cfg_path = Path(args.topcfg).parents[1]
+
+    try:
+        with open(args.topcfg, 'r') as ftop:
+            topcfg = hjson.load(ftop,
+                                use_decimal=True,
+                                object_pairs_hook=OrderedDict)
+    except ValueError:
+        raise SystemExit(sys.exc_info()[1])
+
+
+    # TODO, long term, the levels of dependency should be automatically determined instead
+    # of hardcoded.  The following are a few examples:
+    # Example 1: pinmux depends on amending all modules before calculating the correct number of pins.
+    #            This would be 1 level of dependency and require 2 passes.
+    # Example 2: pinmux depends on amending all modules, and pwrmgr depends on pinmux generation to
+    #            know correct number of wakeups.  This would be 2 levels of dependency and require 3
+    #            passes.
+    process_dependencies = 1
+    for pass_idx in range(process_dependencies+1):
+        log.debug("Generation pass {}".format(pass_idx))
+        if pass_idx < process_dependencies :
+            cfg_copy = deepcopy(topcfg)
+            _process_top(cfg_copy, args, cfg_path, out_path, pass_idx)
+        else:
+            completecfg = _process_top(topcfg, args, cfg_path, out_path, pass_idx)
```

Comment:
I am a little bit confused here. the code updates `cfg_copy` and discard?  I thought the second pass to use the result of amended cfgs. Could you please comment?

Reply:
yeah... so .. our amend updates `topcfg`, and if this is fed back again, it actually creates more warnings / errors (didn't look into it) because the original structure has changed.

So instead what I'm doing is i keep discarding until the last run, where i actually take the results.  The previous runs just generate the correct modules.

So if you take my example..

during the first run we gather all the modules, and generate the correct pinmux.  However, the gathering used stale pinmux, so it goes back again.  

in teh second example, the first run we gather all the modules, and generate the correct pinmux.  However the pwrmgr generation is wrong.  So we go back and second run and gather correct modules including pinmux, and now we generate correct pwrmgr.  Now we go back a 3rd time and use all the correct modules for the final run.

Reply:
AH.. OK Now I got the point! Thanks for explanation. I think it is worth to add your comment to the comment. (hum weird sentence but... you got the point right? :) )

LGTM

Reply:
haha sounds good. It's like a recursive comment!